### ISSUE TITLE
 Kotlin Jackson Mapper not used when on classpath when serializing/deserializing @JdbcTypeCode(SqlTypes.JSON) using Spring Data JPA

### ISSUE THREAD

**Issue by BernhardPosselt** [spring-projects/spring-boot#33870]
After migrating to Spring Boot 3.0.1 and Hibernate 6, I swapped out the Hibernate Types JSON support for the new built in **@JdbcTypeCode(SqlTypes.JSON)**, however my Kotlin data classes failed to deserialize and I had to annotate everything with **@JsonCreator** and **@JsonProperty** to fix constructors.

I feel like this should default to the Jackson instance on the Spring context rather than the Jackson library itself, that is pulled in by Hibernate by default. It should be possible to define a custom mapper factory that can be configured via **hibernate.type.json_format_mapper**  and point Hibernate to that instance, see https://github.com/hibernate/hibernate-orm/blob/main/documentation/src/main/asciidoc/userguide/chapters/domain/basic_types.adoc#json-mapping

**Comment by wilkinsona** [comment]
Thanks for the suggestion. If I have undertood it correctly, you should be able to do this yourself using a `HibernatePropertiesCustomizer` bean:

```java
@Bean
HibernatePropertiesCustomizer jsonFormatMapperCustomizer(ObjectMapper objectMapper) {
	return (properties) -> properties.put(AvailableSettings.JSON_FORMAT_MAPPER,
			new JacksonJsonFormatMapper(objectMapper));
}
```

Can you please try this and let us know if it allows you to remove the `@JsonCreator` and `@JsonProperty` annotations?

**Comment by BernhardPosselt** [comment]
Thank you for getting back, can confirm that this does in fact work fine after removing the annotations.

**Comment by davidfortin55555** [comment]
I have implemented the following code but it is nor working with springboot 3.0.1 and hibernate 6
@Bean
HibernatePropertiesCustomizer jsonFormatMapperCustomizer(ObjectMapper objectMapper) {
	return (properties) -> properties.put(AvailableSettings.JSON_FORMAT_MAPPER,
			new JacksonJsonFormatMapper(objectMapper));
}
Do you habe another proposal?

**Comment by wilkinsona** [comment]
It should work so, without knowing exactly how it isn't working, it's difficult to propose something else. If you'd like some help, please post a question on Stack Overflow that includes a minimal, reproducible example of the problem.

**Comment by wilkinsona** [comment]
While looking at #42676 and trying to find some inspiration for a property name, a couple of other things came to mind:

1. We should consider auto-configuring a `JsonBJsonFormatMapper` when there's a `Jsonb` bean. This also raises a question about what to do when there's both an `ObjectMapper` bean and a `Jsonb` bean. We've got `spring.mvc.converters.preferred-json-mapper` but its name clearly indicates that it shouldn't apply outside of Spring MVC's message conversion.
2. The functionality is marked as `@Incubating` in Hibernate. Are we jumping the gun a bit here and would we be better waiting until its graduated from incubation before we try to offer auto-configuration.

**Comment by philwebb** [comment]
We discussed this today and we'd like to wait until the Hibernate feature is not `@Incubating`.

**Comment by OrangeDog** [comment]
I started looking at Boot 4 today, and it looks like Hibernate 7.1 does not support Jackson 3.

Unlike other dependencies, this did not disqualify it from being supported by Boot, but presumably any user will need to either be stuck on `spring-boot-jackson2`, or hope that a custom `JacksonJsonFormatMapper` works without hitting any other incompatibilities in Hibernate.

**Comment by wilkinsona** [comment]
> Unlike other dependencies, this did not disqualify it from being supported by Boot

No other dependency has been disqualified purely because it does not support Jackson 3. Off the top of my head, Flyway and the Elasticsearch Java client both only support Jackson 2 and have not been disqualified.

> presumably any user will need to either be stuck on spring-boot-jackson2

Nobody needs to be stuck on `spring-boot-jackson2` as it can be used alongside `spring-boot-jackson` during the ecosystem's transitionary period while Jackson 3 support is not universal.

> or hope that a custom JacksonJsonFormatMapper works without hitting any other incompatibilities in Hibernate

I'm not sure what other incompatibilities you're expecting to hit but there should be nothing stopping you from configuring Jackson with a `JacksonJsonFormatMapper` backed by a Jackson 2 `ObjectMapper`. That `ObjectMapper` can be one that you create and configure yourself or, if you add a dependency on spring-boot-jackson2, that is auto-configured by Boot.

**Comment by bclozel** [comment]
@OrangeDog see https://github.com/spring-projects/spring-boot/issues/47789#issuecomment-3443848646 for support in your app and the relevant issue in Hibernate.

**Comment by centfox** [comment]
So apparently, springboot still doesn't support jsonb?

**Comment by bclozel** [comment]
@centfox [We do](https://docs.spring.io/spring-boot/reference/features/json.html#features.json.json-b). I don't think this is related to this issue though.

**Comment by OrangeDog** [comment]
@centfo

If you mean JSON-B the JEE API, do the same as https://github.com/spring-projects/spring-boot/issues/33870#issuecomment-1386822021 but with `JsonBJsonFormatMapper`.

If you mean `jsonb` the SQL data type, just put `@JdbcTypeCode(SqlTypes.JSON)` on the relevant fields.

Both have been supported by Spring for some time.

**Comment by centfox** [comment]
Thank you @OrangeDog for clarification. 
I mean Postgres datatype only. When I use `import tools.jackson.databind.JsonNode` and `@JdbcTypeCode(SqlTypes.JSON) private JsonNode data`,  in my entity, it fails with `com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `tools.jackson.databind.JsonNode` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled)`

I tried in a fresh Springboot 4 project, to be clear, as below.

```
			JsonTest test = new JsonTest();
			test.setId(UUID.randomUUID().toString());
			test.setData(jsonMapper.valueToTree(new String[]{"a","b","c"}));
			dataRepository.save(test);
```
Or is there any specific configuration to be applied?

**Comment by bclozel** [comment]
@centfox you need to configure Jackson 3 support in Hibernate manually because it's not available right now, see https://github.com/spring-projects/spring-boot/issues/47789#issuecomment-3443848646

**Comment by shobhitpharmeasy** [comment]
How could we use this with spring boot 4? tools.jackson.objectmapper module?

**Comment by wilkinsona** [comment]
Hibernate has added [Jackson 3 support](https://docs.hibernate.org/orm/7.3/whats-new/#jackson-3-support) in 7.3. Spring Boot does not yet support 7.3 due to [a regression](https://hibernate.atlassian.net/browse/HHH-20265) that causes a StackOverflowError when introspecting entities. You may want to try a manual upgrade to Hibernate 7.3 to see if you are affected. If you are, please upvote or comment on the Hibernate issue as appropriate.

**Comment by BernhardPosselt** [comment]
Will this prevent 7.3 from being shipped in 4.1?

**Comment by wilkinsona** [comment]
Yes. Without a fix in Hibernate we can't upgrade to 7.3.x as it breaks any application using Spring Data JPA's `AbstractAuditable`. As we understand it, any other entity hierarchy that makes similar use of generics is also affected.

**Comment by dmitrysulman** [comment]
@wilkinsona Since Hibernate [has been upgraded to 7.4](https://github.com/spring-projects/spring-boot/issues/50578), do you think there's a chance this could be shipped in Spring Boot 4.1?

**Comment by wilkinsona** [comment]
No. `FormatMapper` is still `@Incubating` so this issue remains [blocked](https://github.com/spring-projects/spring-boot/issues/33870#issuecomment-2573476473).

**Comment by OrangeDog** [comment]
In case people missed it, everything works fine if you just do it yourself:
```java
public class Jackson3FormatMapper extends AbstractJsonFormatMapper {
    private final ObjectMapper objectMapper;
    public Jackson3FormatMapper(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }
    protected <T> String toString(T t, Type type) {
        return objectMapper.writerFor(objectMapper.constructType(type)).writeValueAsString(t);
    }
    protected <T> T fromString(CharSequence charSequence, Type type) {
        return objectMapper.readerFor(objectMapper.constructType(type)).readValue(charSequence.toString());
    }
    public boolean supportsSourceType(Class<?> sourceType) {
        return JsonParser.class.isAssignableFrom(sourceType);
    }
    public boolean supportsTargetType(Class<?> targetType) {
        return JsonGenerator.class.isAssignableFrom(targetType);
    }
    public <T> void writeToTarget(T value, JavaType<T> javaType, Object target, WrapperOptions options) throws IOException {
        objectMapper.writerFor(objectMapper.constructType(javaType.getJavaType())).writeValue((JsonGenerator) target, value);
    }
    public <T> T readFromSource(JavaType<T> javaType, Object source, WrapperOptions options) throws IOException {
        return objectMapper.readerFor(objectMapper.constructType(javaType.getJavaType())).readValue((JsonParser) source);
    }
}

@Bean
public HibernatePropertiesCustomizer hibernatePropertiesCustomizer(ObjectMapper objectMapper) {
    return properties -> properties.put(
            AvailableSettings.JSON_FORMAT_MAPPER, new Jackson3FormatMapper(objectMapper)
    );
}
```

All that's missing is Boot doing it automatically for you.

**Comment by radistao** [comment]
Tried on Spring Boot 4.1.0, Jakarta persistence 3.2.0, Hibernate 7.4.1:
* by default: if there are both Jackson 2 and Jackson 3 mappers in the classpath - Jackson 2 has priority (see [SessionFactoryOptionsBuilder](https://github.com/hibernate/hibernate-orm/blob/7.4/hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java#L913-L920)
* jackson 3 could be prioritized by configuring either (they are equal):
  * `spring.jpa.properties.hibernate.type.json_format_mapper=jackson3`
  *  FQDN `spring.jpa.properties.hibernate.type.json_format_mapper=org.hibernate.type.format.jackson.Jackson3JsonFormatMapper`
Note ℹ️ : default Hibernate Jackson3 mapper builder falls back to <code>builderWithJackson<b>2</b>Defaults</code>

Specifying custom mapper possible either by hibernate runtime configuration [discussed above](https://github.com/spring-projects/spring-boot/issues/33870#issuecomment-1386822021) or configuring own one, extending `org.hibernate.type.format.FormatMapper`: 
`spring.jpa.properties.hibernate.type.json_format_mapper=com.example.MyHibernateJacksonFormatMapper`

Example (Kotlin):
<details>

<summary>Custom Hibernate Jackson Mapper</summary>

```kotlin
class MyHibernateJacksonFormatMapper : AbstractJsonFormatMapper() {

    override fun <T> fromString(
        charSequence: CharSequence?,
        type: Type?,
    ): T? = hibernateJackson3JsonFormatMapper.fromString(charSequence, type)

    override fun <T> toString(
        value: T?,
        type: Type?,
    ): String? = hibernateJackson3JsonFormatMapper.toString(value, type)

    private companion object {

        // could be reused in other places
        private val customObjectDbMapper: JsonMapper by lazy {
            jsonMapper {
                addModule(
                    kotlinModule {
                        enable(KotlinFeature.NullIsSameAsDefault)
                        // this was disabled by default in Jackson 2, but required for legacy DB entities mapping
                        disable(KotlinFeature.StrictNullChecks)
                    },
                )

                enable(StreamWriteFeature.WRITE_BIGDECIMAL_AS_PLAIN) // avoid using scientific notation

                disable(DeserializationFeature.FAIL_ON_NULL_CREATOR_PROPERTIES)
                disable(DateTimeFeature.WRITE_DATES_AS_TIMESTAMPS) // timestamps should be serialized using default timestamps

                // this was disabled by default in Jackson 2, but required for legacy DB entities mapping (when new properties were added)
                disable(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES)

                enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
                disable(DateTimeFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE)
            }
        }

        private val hibernateJackson3JsonFormatMapper by lazy { Jackson3JsonFormatMapper(customObjectDbMapper) }
    }
}

```

</details>

All three proved to be working like a charm.

**Comment by OrangeDog** [comment]
@radistao the actual equivalent to what this issue wants would be:

```java
@Bean
public HibernatePropertiesCustomizer hibernatePropertiesCustomizer(JsonMapper jsonMapper) {
    return properties -> properties.put(
            AvailableSettings.JSON_FORMAT_MAPPER, new Jackson3JsonFormatMapper(objectMapper)
    );
}
```
